File: core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java
package fr.sncf.osrd.stdcm.preprocessing.implementation;

import static fr.sncf.osrd.utils.KtToJavaConverter.toIntList;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.Multimap;
import com.google.common.collect.Range;
import com.google.common.collect.RangeSet;
import com.google.common.collect.TreeRangeSet;
import fr.sncf.osrd.envelope_sim.PhysicsRollingStock;
import fr.sncf.osrd.sim_infra.api.BlockInfra;
import fr.sncf.osrd.sim_infra.api.RawSignalingInfra;
import fr.sncf.osrd.sim_infra.impl.BlockInfraImplKt;
import fr.sncf.osrd.standalone_sim.result.ResultTrain;
import fr.sncf.osrd.stdcm.OccupancySegment;
import fr.sncf.osrd.utils.units.Distance;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class UnavailableSpaceBuilder {
    private static final long SIGHT_DISTANCE = Distance.fromMeters(400);

    /** Computes the unavailable space for each block, i.e.
     * the times and positions where the *head* of the train cannot be.
     * This considers existing occupancy segments, the length of the train,
     * and the blocks that must be left available behind the train
     * <br/>
     * This is the first step to compute STDCM, the goal is to get rid of railway rules and extra complexity
     * as soon as possible. After this step we can look for a single curve that avoids unavailable segment. */
    public static Multimap<Integer, OccupancySegment> computeUnavailableSpace(
            RawSignalingInfra infra,
            BlockInfra blockInfra,
            Collection<ResultTrain.SpacingRequirement> spacingRequirements,
            PhysicsRollingStock rollingStock,
            double marginToAddBeforeEachBlock,
            double marginToAddAfterEachBlock
    ) {
        Multimap<Integer, OccupancySegment> unavailableSpace = HashMultimap.create();

        var blockUse = buildBlockUse(infra, blockInfra, spacingRequirements,
                marginToAddBeforeEachBlock, marginToAddAfterEachBlock);

        for (var blockEntry : blockUse.entrySet()) {
            var blockId = blockEntry.getKey();
            var blockLength = blockInfra.getBlockLength(blockId);
            var useTimes = blockEntry.getValue();
            for (var timeRange : useTimes.asRanges()) {

                // Generate current block occupancy
                unavailableSpace.put(blockId, new OccupancySegment(
                        timeRange.lowerEndpoint(), timeRange.upperEndpoint(), 0, blockLength)
                );

                // Generate the warnings in blocks before the ones used by other trains
                var predecessorBlocks = getPreviousBlocks(infra, blockInfra, blockId);
                for (var predecessorBlock : predecessorBlocks) {
                    var preBlockLength = blockInfra.getBlockLength(predecessorBlock);
                    unavailableSpace.put(predecessorBlock, new OccupancySegment(
                            timeRange.lowerEndpoint(), timeRange.upperEndpoint(), 0, preBlockLength
                    ));

                    // Generate the sight distance requirements in the blocks before that
                    for (var secondPredecessorBlock : getPreviousBlocks(infra, blockInfra, predecessorBlock)) {
                        var secPreBlockLength = blockInfra.getBlockLength(secondPredecessorBlock);
                        unavailableSpace.put(secondPredecessorBlock, new OccupancySegment(
                                timeRange.lowerEndpoint(), timeRange.upperEndpoint(),
                                Math.max(0, secPreBlockLength - SIGHT_DISTANCE), preBlockLength
                        ));
                    }
                }

                // Generate train length occupancy
                var successorBlocks = getNextBlocks(infra, blockInfra, blockId);
                for (var successorBlock : successorBlocks) {
                    var nextBlockLength = blockInfra.getBlockLength(successorBlock);
                    unavailableSpace.put(successorBlock, new OccupancySegment(
                            timeRange.lowerEndpoint(), timeRange.upperEndpoint(),
                            0, Math.min(nextBlockLength, Distance.fromMeters(rollingStock.getLength()))
                    ));
                }
            }
        }
        return unavailableSpace;
    }

    /** Builds the time during which the blocks are used by another train (including a forward signal cascade)
     * This step is also used to merge together the small overlapping time intervals
     * across different zones or trains. */
    private static Map<Integer, RangeSet<Double>> buildBlockUse(
            RawSignalingInfra infra,
            BlockInfra blockInfra,
            Collection<ResultTrain.SpacingRequirement> requirements,
            double marginToAddBeforeEachBlock,
            double marginToAddAfterEachBlock
    ) {
        var res = new HashMap<Integer, RangeSet<Double>>();
        for (var requirement : requirements) {
            var zoneId = infra.getZoneFromName(requirement.zone);
            var timeRange = Range.closed(
                    requirement.beginTime - marginToAddBeforeEachBlock,
                    requirement.endTime + marginToAddAfterEachBlock
            );
            for (var blockId : toIntList(blockInfra.getBlocksInZone(zoneId))) {
                res.putIfAbsent(blockId, TreeRangeSet.create());
                res.get(blockId).add(timeRange);
            }
        }
        return res;
    }

    /** Returns the blocks that lead into the given one */
    private static Set<Integer> getPreviousBlocks(RawSignalingInfra infra, BlockInfra blockInfra, int blockId) {
        var entry = BlockInfraImplKt.getBlockEntry(blockInfra, infra, blockId);
        return new HashSet<>(toIntList(blockInfra.getBlocksEndingAtDetector(entry)));
    }

    /** Returns the blocks that follow the given one */
    private static Set<Integer> getNextBlocks(RawSignalingInfra infra, BlockInfra blockInfra, int blockId) {
        var entry = BlockInfraImplKt.getBlockExit(blockInfra, infra, blockId);
        return new HashSet<>(toIntList(blockInfra.getBlocksStartingAtDetector(entry)));
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.kt
package fr.sncf.osrd.stdcm.preprocessing.implementation

import com.google.common.collect.*
import fr.sncf.osrd.envelope_sim.PhysicsRollingStock
import fr.sncf.osrd.sim_infra.api.BlockId
import fr.sncf.osrd.sim_infra.api.BlockInfra
import fr.sncf.osrd.sim_infra.api.RawSignalingInfra
import fr.sncf.osrd.sim_infra.utils.getNextBlocks
import fr.sncf.osrd.sim_infra.utils.getPreviousBlocks
import fr.sncf.osrd.standalone_sim.result.ResultTrain.SpacingRequirement
import fr.sncf.osrd.stdcm.OccupancySegment
import fr.sncf.osrd.utils.units.Distance.Companion.fromMeters
import fr.sncf.osrd.utils.units.Distance.Companion.max
import fr.sncf.osrd.utils.units.Distance.Companion.min
import fr.sncf.osrd.utils.units.meters

private val SIGHT_DISTANCE = 400.meters

/**
 * Computes the unavailable space for each block, i.e. the times and positions where the *head* of
 * the train cannot be. This considers existing occupancy segments, the length of the train, and the
 * blocks that must be left available behind the train <br></br> This is the first step to compute
 * STDCM, the goal is to get rid of railway rules and extra complexity as soon as possible. After
 * this step we can look for a single curve that avoids unavailable segment.
 */
fun computeUnavailableSpace(
    rawInfra: RawSignalingInfra,
    blockInfra: BlockInfra,
    spacingRequirements: Collection<SpacingRequirement>,
    rollingStock: PhysicsRollingStock,
    marginToAddBeforeEachBlock: Double,
    marginToAddAfterEachBlock: Double
): Multimap<BlockId, OccupancySegment> {
    val unavailableSpace: Multimap<BlockId, OccupancySegment> = HashMultimap.create()
    val blockUse =
        buildBlockUse(
            rawInfra,
            blockInfra,
            spacingRequirements,
            marginToAddBeforeEachBlock,
            marginToAddAfterEachBlock
        )
    for ((blockId, useTimes) in blockUse) {
        val blockLength = blockInfra.getBlockLength(blockId)
        for (timeRange in useTimes.asRanges()) {

            // Generate current block occupancy
            unavailableSpace.put(
                blockId,
                OccupancySegment(
                    timeRange.lowerEndpoint(),
                    timeRange.upperEndpoint(),
                    0.meters,
                    blockLength.distance
                )
            )

            // Generate the warnings in blocks before the ones used by other trains
            val predecessorBlocks = blockInfra.getPreviousBlocks(rawInfra, blockId)
            for (predecessorBlock in predecessorBlocks) {
                val preBlockLength = blockInfra.getBlockLength(predecessorBlock)
                unavailableSpace.put(
                    predecessorBlock,
                    OccupancySegment(
                        timeRange.lowerEndpoint(),
                        timeRange.upperEndpoint(),
                        0.meters,
                        preBlockLength.distance
                    )
                )

                // Generate the sight distance requirements in the blocks before that
                for (secondPredecessorBlock in
                    blockInfra.getPreviousBlocks(rawInfra, predecessorBlock)) {
                    val secPreBlockLength = blockInfra.getBlockLength(secondPredecessorBlock)
                    unavailableSpace.put(
                        secondPredecessorBlock,
                        OccupancySegment(
                            timeRange.lowerEndpoint(),
                            timeRange.upperEndpoint(),
                            max(0.meters, secPreBlockLength.distance - SIGHT_DISTANCE),
                            secPreBlockLength.distance
                        )
                    )
                }
            }

            // Generate train length occupancy
            val successorBlocks = blockInfra.getNextBlocks(rawInfra, blockId)
            for (successorBlock in successorBlocks) {
                val nextBlockLength = blockInfra.getBlockLength(successorBlock)
                unavailableSpace.put(
                    successorBlock,
                    OccupancySegment(
                        timeRange.lowerEndpoint(),
                        timeRange.upperEndpoint(),
                        0.meters,
                        min(nextBlockLength.distance, fromMeters(rollingStock.length))
                    )
                )
            }
        }
    }
    // validation
    for ((key, value) in unavailableSpace.entries()) {
        assert(value.distanceStart <= value.distanceEnd)
        assert(value.timeStart <= value.timeEnd)
        assert(value.distanceEnd <= blockInfra.getBlockLength(key).distance)
        assert(0.meters <= value.distanceStart)
        assert(0.0 <= value.timeEnd)
    }
    return unavailableSpace
}

/**
 * Builds the time during which the blocks are used by another train (including a forward signal
 * cascade) This step is also used to merge together the small overlapping time intervals across
 * different zones or trains.
 */
private fun buildBlockUse(
    rawInfra: RawSignalingInfra,
    blockInfra: BlockInfra,
    requirements: Collection<SpacingRequirement>,
    marginToAddBeforeEachBlock: Double,
    marginToAddAfterEachBlock: Double
): Map<BlockId, RangeSet<Double>> {
    val res = HashMap<BlockId, RangeSet<Double>>()
    for (requirement in requirements) {
        val zoneId = rawInfra.getZoneFromName(requirement.zone)
        val timeRange =
            Range.closed(
                requirement.beginTime - marginToAddBeforeEachBlock,
                requirement.endTime + marginToAddAfterEachBlock
            )
        for (blockId in blockInfra.getBlocksInZone(zoneId)) {
            res.putIfAbsent(blockId, TreeRangeSet.create())
            res[blockId]!!.add(timeRange)
        }
    }
    return res
}
